### PR TITLE
chore: return ErrNotFound when possible

### PR DESCRIPTION
**Pull request by eecsliu**
## Description
Bun looks like it bubbles up sql errors instead of wrapping its own, which makes sense but makes this a bit harder to do. Of the errors we've defined in database.go, `ErrNotFound` is the only one that I think we can easily identify. 

## Checklist

- [ ] User-facing API changes need the "User-facing API Change" label.
- [ ] Release notes should be added as a separate file under `docs/release-notes/`.
See [Release Note](../docs/release-notes/README.md) for details.
- [ ] Licenses should be included for new code which was copied and/or modified from any external code.

### CODE REVIEW

File: master/internal/db/postgres_groups.go [hunk 1]
```diff
@@ -30,22 +35,28 @@ func (db *PgDB) SearchGroups(ctx context.Context, userBelongsTo model.UserID) ([
 	}
 
 	err := query.Scan(ctx)
+
+	if len(groups) < 1 || errors.Is(err, sql.ErrNoRows) {
```
Comment (maxrussell) [inline master/internal/db/postgres_groups.go]: This is a judgment call, but I think we shouldn't return `ErrNotFound` when we search for groups or users but find nothing. In my opinion, it's not necessarily an error condition when we look for a set of things and the set is empty. By analogy, if I ran a command like this:

```
find / -name 'some-weird-file-name.go'
```

and had no file in my filesystem with that name, I'd expect the `find` command to exit with code 0 rather than 1; it successfully ran its search, and returned correct results: that the set of matches was empty.

Comment (eecsliu) [reply master/internal/db/postgres_groups.go]: yeah that makes sense - I changed this to return whatever the query returned

File: master/internal/db/postgres_groups.go [hunk 2]
```diff
@@ -57,17 +68,18 @@ func (db *PgDB) AddUsersToGroup(ctx context.Context, gid int, uids ...model.User
 		})
 	}
 
-	_, err := Bun().NewInsert().Model(&groupMem).Exec(ctx)
-	return err
+	res, err := Bun().NewInsert().Model(&groupMem).Exec(ctx)
+
+	return checkIfFound(res, err)
```
Comment (maxrussell) [inline master/internal/db/postgres_groups.go:80]: Does this return `ErrorNotFound` when the group and/or any of the users is not found? That makes me think of an edge case we should probably test: if this operation fails, we want the system to behave in an intended, predictable way—either rolling back the whole transaction or keeping around the few that did succeed. Either way, we should have a test for this behavior once we decide how it should work. In my opinion, the whole transaction should succeed or fail. What do you think?

File: master/internal/db/postgres_groups.go [hunk 3]
```diff
@@ -77,5 +89,22 @@ func (db *PgDB) GetUsersInGroup(ctx context.Context, gid int) ([]model.User, err
 		Where("ugm.group_id = ?", gid).
 		Scan(ctx)
 
+	if len(users) < 1 || errors.Is(err, sql.ErrNoRows) {
+		return users, ErrNotFound
+	}
```
Comment (maxrussell) [inline master/internal/db/postgres_groups.go]: Same as above: I think this search should return no error if we successfully search but the set is empty.

File: master/internal/db/postgres_groups.go [hunk 4]
```diff
@@ -77,5 +89,22 @@ func (db *PgDB) GetUsersInGroup(ctx context.Context, gid int) ([]model.User, err
 		Where("ugm.group_id = ?", gid).
 		Scan(ctx)
 
+	if len(users) < 1 || errors.Is(err, sql.ErrNoRows) {
```
Comment (maxrussell) [inline master/internal/db/postgres_groups.go]: Same as above: I think this search should return no error if we successfully search but the set is empty.

File: master/internal/db/postgres_groups.go (hunk 2 above)
Comment (eecsliu) [reply master/internal/db/postgres_groups.go:80]: At a cursory look, it seems like bun provides a transaction abstraction that we might be able to take advantage of. I think if we wrap some things in a transaction we can automatically rollback in the event of an error?

File: master/internal/db/postgres_groups.go [hunk 5]
```diff
@@ -57,17 +65,22 @@ func (db *PgDB) AddUsersToGroup(ctx context.Context, gid int, uids ...model.User
 		})
 	}
 
-	_, err := Bun().NewInsert().Model(&groupMem).Exec(ctx)
-	return err
+	tx, err := Bun().BeginTx(ctx, &sql.TxOptions{})
```
Comment (maxrussell) [inline master/internal/db/postgres_groups.go:68]: Nice! Thanks for looking into this and taking care of it.

We'll wanna check this error instead of overwriting it on the next line.

It's also really important with database transactions not to leave them "hanging." For example, if there was a `panic` before the `tx.Commit()`, the transaction would be left open, which can severely impact database performance or even lock a table or row (depending on DBMS, queries run in the transaction, and/or settings). Since `tx.Rollback()` is a no-op on an already committed transaction, a common idiom in Go is to `defer tx.Rollback()` immediately after successfully beginning. I think we'll want something like this:

```go
tx, err := Bun().BeginTx(ctx, &sql.TxOptions{})
if err != nil {
	return err
}
defer tx.Rollback()

res, err := tx.NewInsert().Model(&groupMem).Exec(ctx)
if err != nil {
	return checkIfFound(res, err)
}

err = tx.Commit()
return checkIfFound(res, err)
```

Comment (maxrussell) [reply master/internal/db/postgres_groups.go:68]: Another option is to use Bun's `db.RunInTx` feature. It's 100% personal taste, and totally fine if you prefer either, but I don't like `RunInTx` as much.
